pkg/cloud/services/secretsmanager/secret_test.go

```diff
@@ -0,0 +1,291 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package secretsmanager
+
+import (
+	"math/rand"
+	"sort"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/awserrors"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/secretsmanager/mock_secretsmanageriface"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+func TestService_Create(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	generateBytes := func(g *WithT, count int64) []byte {
+		token := make([]byte, count)
+		_, err := rand.Read(token)
+		g.Expect(err).NotTo(HaveOccurred())
+		return token
+	}
+
+	sortTagsByKey := func(tags []*secretsmanager.Tag) {
+		sort.Slice(tags, func(i, j int) bool {
+			return *(tags[i].Key) < *(tags[j].Key)
+		})
+	}
+
+	expectedTags := []*secretsmanager.Tag{
+		{
+			Key:   aws.String("Name"),
+			Value: aws.String("infra-cluster"),
+		},
+		{
+			Key:   aws.String("kubernetes.io/cluster/test"),
+			Value: aws.String("owned"),
+		},
+		{
+			Key:   aws.String("sigs.k8s.io/cluster-api-provider-aws/cluster/test"),
+			Value: aws.String("owned"),
+		},
+		{
+			Key:   aws.String("sigs.k8s.io/cluster-api-provider-aws/role"),
+			Value: aws.String("node"),
+		},
+	}
+
+	check := func(g *WithT, actualPrefix string, expectedPrefix string, err error, wantErr bool) {
+		g.Expect(actualPrefix).Should(HavePrefix(expectedPrefix))
+		if wantErr {
+			g.Expect(err).To(HaveOccurred())
+			return
+		}
+		g.Expect(err).NotTo(HaveOccurred())
+	}
+
+	tests := []struct {
+		name           string
+		bytesCount     int64
+		secretPrefix   string
+		expectedPrefix string
+		wantErr        bool
+		expect         func(g *WithT, m *mock_secretsmanageriface.MockSecretsManagerAPIMockRecorder)
+	}{
+		{
+			name:           "Should not store data in secret manager if data is having zero bytes",
```

Comment:
How is this scenario checked here by just checking nil error?

Reply:
The expect method of this test is nil, which means this test will succeed if there is no call to CreateSecret() method. If there is any call to CreateSecret(), the test will fail.
For other tests where we are expecting to store data in secret manager, we have set expected calls to CreateSecret() method.

I will add an expect method having `m.CreateSecret(gomock.Any()).Times(0)` to this test to make this more explicit.

Reply:
No, this is enough. Thanks.